modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C

```diff
@@ -0,0 +1,89 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PorousFlowEnthalpySink.h"
+#include "Function.h"
+#include "PorousFlowDictator.h"
+#include "SinglePhaseFluidProperties.h"
+#include "MooseVariable.h"
+
+#include "libmesh/quadrature.h"
+
+#include <iostream>
+
+registerMooseObject("PorousFlowApp", PorousFlowEnthalpySink);
+
+InputParameters
+PorousFlowEnthalpySink::validParams()
+{
+  InputParameters params = IntegratedBC::validParams();
+  params.addRequiredParam<UserObjectName>(
+      "PorousFlowDictator", "The UserObject that holds the list of PorousFlow variable names");
+  params.addParam<unsigned int>(
+      "fluid_phase", "If supplied, then this BC will potentially be a function of fluid pressure.");
+  params.addRequiredParam<FunctionName>(
+      "flux_function",
+      "The mass flux.  The flux is OUT of the medium: hence positive values of "
+      "this function means this BC will act as a SINK, while negative values "
+      "indicate this flux will be a SOURCE.  The functional form is useful "
+      "for spatially or temporally varying sinks.");
+  params.addRequiredParam<UserObjectName>("fp", "The name of the user object for fluid properties");
+  params.addRequiredCoupledVar("pressure", "Pressure");
+  params.addRequiredParam<Real>("T_in", "Specified inlet temperature");
+  params.addClassDescription("Applies a flux sink to a boundary.");
+
+  return params;
+}
+
+PorousFlowEnthalpySink::PorousFlowEnthalpySink(const InputParameters & parameters)
+  : IntegratedBC(parameters),
+    _dictator(getUserObject<PorousFlowDictator>("PorousFlowDictator")),
+    _involves_fluid(isParamValid("fluid_phase")),
+    _ph(_involves_fluid ? getParam<unsigned int>("fluid_phase") : 0),
+    _m_func(getFunction("flux_function")),
+    _pp(getMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_nodal")),
+    _dpp_dvar(
+        getMaterialProperty<std::vector<std::vector<Real>>>("dPorousFlow_porepressure_nodal_dvar")),
+    _T_in(getParam<Real>("T_in")),
+    _fp(getUserObject<SinglePhaseFluidProperties>("fp"))
+{
+}
+
+Real
+PorousFlowEnthalpySink::computeQpResidual()
+{
+  Real h = _fp.h_from_p_T(_pp[_qp][_ph], _T_in);
```

Comment:
Please think about this: i believe you should have `_pp[_i][_ph]` because we probably want to use the upwinded version.  In your case i don't think it matters hugely since this is a source and not a sink (upwinding is especially important in sinks, because quadpoint stuff can inadvertently pull out too much fluid/heat resulting in negative fluid mass or temperature).  Also, you don't need to upwind any mobilities, etc.  But everywhere else we've used full upwinding, so i suggest to use it here.  If you want to use `_pp[_qp][_ph]` then you should pull in "PorousFlow_porepressure_qp" instead of "PorousFlow_porepressure_nodal" (and same for the derivative).

Then your code will be like line238 of `PorousFlowSink.C`.

Reply:
I think you are right. My original version was using just qp-based values and this did not get updated when I pulled in `PorousFlow_porepressure_nodal`. Still learning your code ;-)